conjure-java-core/src/test/java/com/palantir/conjure/java/types/Union2.java

```diff
@@ -0,0 +1,215 @@
+     */
+    @Deprecated
+    static Union2 baz(long value) {
+        return new Baz(value);
+    }
+
+    static Union2 unknown(@Safe String type, Object value) {
+        return switch (Preconditions.checkNotNull(type, "Type is required")) {
+            case "foo" -> throw new SafeIllegalArgumentException(
+                    "Unknown type cannot be created as the provided type is known: foo");
+            case "bar" -> throw new SafeIllegalArgumentException(
+                    "Unknown type cannot be created as the provided type is known: bar");
+            case "baz" -> throw new SafeIllegalArgumentException(
+                    "Unknown type cannot be created as the provided type is known: baz");
+            default -> new UnknownVariant(type, Collections.singletonMap(type, value));
+        };
+    }
+
+    default Known throwOnUnknown() {
+        if (this instanceof UnknownVariant) {
+            throw new SafeIllegalArgumentException(
+                    "Unknown variant of the 'Union' union",
+                    SafeArg.of("unknownType", ((UnknownVariant) this).getType()));
+        } else {
+            return (Known) this;
+        }
+    }
+
+    @JsonTypeName("foo")
+    record Foo(String value) implements Union2, Known {
+        @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+        public Foo(@JsonSetter("foo") @Nonnull String value) {
+            Preconditions.checkNotNull(value, "foo cannot be null");
+            this.value = value;
+        }
+
+        @JsonProperty(value = "type", index = 0)
+        @SuppressWarnings("UnusedMethod")
+        private String getType() {
+            return "foo";
+        }
+
+        @JsonProperty("foo")
+        public String getValue() {
+            return value;
+        }
+    }
+
+    @JsonTypeName("bar")
+    record Bar(int value) implements Union2, Known {
+        @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+        public Bar(@JsonSetter("bar") @Nonnull int value) {
+            Preconditions.checkNotNull(value, "bar cannot be null");
+            this.value = value;
+        }
+
+        @JsonProperty(value = "type", index = 0)
+        @SuppressWarnings("UnusedMethod")
+        private String getType() {
+            return "bar";
+        }
+
+        @JsonProperty("bar")
+        public int getValue() {
+            return value;
+        }
+    }
+
+    @JsonTypeName("baz")
+    record Baz(long value) implements Union2, Known {
+        @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+        public Baz(@JsonSetter("baz") @Nonnull long value) {
+            Preconditions.checkNotNull(value, "baz cannot be null");
+            this.value = value;
+        }
+
+        @JsonProperty(value = "type", index = 0)
+        @SuppressWarnings("UnusedMethod")
+        private String getType() {
+            return "baz";
+        }
+
+        @JsonProperty("baz")
+        public long getValue() {
+            return value;
+        }
+    }
```

Comment:
Entirely untested, but I wonder if can do something like this:
```suggestion
    @JsonTypeName("baz")
    @JsonAppend(attrs = @JsonAppend.Attr(propName = "type", value = "baz"), prepend = true)
    record Baz(long value) implements Union2, Known {
        @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
        public Baz(@JsonSetter("baz") @Nonnull long value) {
            Preconditions.checkNotNull(value, "baz cannot be null");
            this.value = value;
        }

        @JsonProperty("baz")
        public long getValue() {
            return value;
        }
    }
```

Reply:
Diff didn't render how I expected... The above suggests adding an annotation to the record: `@JsonAppend(attrs = @JsonAppend.Attr(propName = "type", value = "baz"), prepend = true)` and removing the `getType()` method.